File: backend/app/api/models/screening.py
"""
MedAI Hub - Screening Models
Pydantic models for the Smart Screener module.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

# ----------------------------------------------------------------------------
# Criteria Configuration
# ----------------------------------------------------------------------------


class PopulationCriteria(BaseModel):
    codes: list[str] = Field(
        default_factory=list, description="Selected population codes (e.g., 'P1')"
    )
    custom_text: str | None = Field(None, description="Free text description for custom population")


class StudyDesignCriteria(BaseModel):
    inclusion_codes: list[str] = Field(
        default_factory=list, description="Codes to include (e.g., 'S2')"
    )
    exclusion_codes: list[str] = Field(
        default_factory=list, description="Codes to exclude (e.g., 'S-Ex1')"
    )


class CriteriaConfig(BaseModel):
    """
    Configuration for a screening run.
    Snapshot of settings used to filter articles.
    """

    review_type: str = Field(..., description="systematic, scoping, or quick")
    date_range_start: int | None = Field(None, description="Start year (e.g. 2015)")
    date_range_end: int | None = Field(None, description="End year (e.g. 2025)")
    languages: list[str] = Field(default=["eng"], description="Allowed languages")

    population: PopulationCriteria = Field(default_factory=PopulationCriteria)
    study_design: StudyDesignCriteria = Field(default_factory=StudyDesignCriteria)

    # Custom free-text criteria for AI to consider
    custom_inclusion: str | None = None
    custom_exclusion: str | None = None


class ScreeningCriteriaCreate(BaseModel):
    project_id: UUID
    review_type: str
    filters: CriteriaConfig


class ScreeningCriteriaResponse(ScreeningCriteriaCreate):
    id: UUID
    created_at: datetime

    class Config:
        from_attributes = True


# ----------------------------------------------------------------------------
# Article Decisions
# ----------------------------------------------------------------------------


class ArticleDecisionCreate(BaseModel):
    project_id: UUID
    pmid: str
    title: str | None = None

    # Decision
    source: str = Field(..., description="rule_engine, ai_model, human_override")
    status: str = Field(..., description="included, excluded, unclear")
    reason: str | None = None

    # AI Metadata
    evidence_quote: str | None = None
    study_type_classification: str | None = None


class ArticleDecisionUpdate(BaseModel):
    human_override_status: str | None = None
    human_notes: str | None = None


class ArticleDecisionResponse(ArticleDecisionCreate):
    human_override_status: str | None = None
    human_notes: str | None = None
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/types/screening.ts
/**
 * Type definitions for Smart Screener module
 */

// ============================================================================
// Criteria Library Types
// ============================================================================

export interface CriteriaCode {
  code: string;
  label: string;
  description: string;
  exclude: boolean;
  query_filter?: string;
}

export interface CriteriaLibrary {
  population: Record<string, CriteriaCode>;
  intervention: Record<string, CriteriaCode>;
  comparator: Record<string, CriteriaCode>;
  outcome: Record<string, CriteriaCode>;
  study_design: Record<string, CriteriaCode>;
}

// ============================================================================
// Criteria Configuration Types
// ============================================================================

export interface PopulationCriteria {
  codes: string[];
  custom_text?: string;
}

export interface StudyDesignCriteria {
  inclusion_codes: string[];
  exclusion_codes: string[];
}

export interface CriteriaConfig {
  review_type: 'systematic' | 'scoping' | 'quick';
  date_range_start?: number;
  date_range_end?: number;
  languages: string[];
  population: PopulationCriteria;
  study_design: StudyDesignCriteria;
  custom_inclusion?: string;
  custom_exclusion?: string;
}

// ============================================================================
// Request/Response Types
// ============================================================================

export interface InitCriteriaRequest {
  project_id: string;
  review_type: string;
  filters: CriteriaConfig;
}

export interface InitCriteriaResponse {
  id: string;
  project_id: string;
  review_type: string;
  filters: CriteriaConfig;
  created_at: string;
}

export interface ProcessPmidsRequest {
  project_id: string;
  pmids: string[];
  criteria_config: CriteriaConfig;
}

export interface ArticleDecision {
  pmid: string;
  title: string;
  status: 'included' | 'excluded' | 'unclear';
  source: 'rule_engine' | 'ai_model';
  reason: string;
  evidence_quote?: string;
  study_type_classification?: string;
}

export interface ScreeningResult {
  processed: number;
  included: number;
  excluded: number;
  details: ArticleDecision[];
}

// ============================================================================
// UI State Types
// ============================================================================

export interface WizardState {
  currentStep: number;
  projectId: string;
  frameworkData?: Record<string, string>;
  reviewType?: 'systematic' | 'scoping' | 'quick';
  criteriaConfig?: CriteriaConfig;
  pmids?: string[];
  result?: ScreeningResult;
}

export interface ReviewTypeInfo {
  id: 'systematic' | 'scoping' | 'quick';
  name: string;
  description: string;
  icon: string;
  characteristics: string[];
  recommended_for: string;
}
